### PR TITLE
Make CI happy again

### PR DESCRIPTION
**Pull request by milesgranger**
xref: https://github.com/dask-contrib/dask-expr/issues/945

Tried to split each 'fix' into it's own commit, I think some may need their own follow-up issue.

### CODE REVIEW

File: distributed/protocol/tests/test_highlevelgraph.py [hunk 1]
```diff
@@ -155,6 +155,7 @@ def fn(k):
     assert plugin.resource_matches == B.npartitions
 
 
+@pytest.mark.xfail(dd._dask_expr_enabled(), reason="Annotation is WIP in dask-expr")
 @gen_cluster(client=True)
 async def test_dataframe_annotations(c, s, a, b):
```
Comment (fjetter) [inline distributed/protocol/tests/test_highlevelgraph.py:160]: see also https://github.com/dask-contrib/dask-expr/pull/947

File: distributed/shuffle/tests/test_graph.py [hunk 2]
```diff
@@ -37,10 +37,16 @@ def test_raise_on_complex_numbers(dtype):
...
         pd.DataFrame({"x": pd.array(range(10), dtype=dtype)}), npartitions=5
     )
-    with pytest.raises(
-        TypeError, match=f"p2p does not support data of type '{df.x.dtype}'"
-    ), dask.config.set({"dataframe.shuffle.method": "p2p"}):
-        df.shuffle("x")
+
+    if dd._dask_expr_enabled():
+        np = pytest.importorskip("numpy")
+        with pytest.raises(np.exceptions.ComplexWarning):
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_graph.py:43]: We should fix this, having a numpy warning bubble up isn't great UX.

File: distributed/shuffle/tests/test_graph.py [hunk 3]
```diff
@@ -65,18 +71,25 @@ def test_raise_on_sparse_data():
     df = dd.from_pandas(
         pd.DataFrame({"x": pd.array(range(10), dtype="Sparse[float64]")}), npartitions=5
     )
-    with pytest.raises(
-        TypeError, match="p2p does not support sparse data"
-    ), dask.config.set({"dataframe.shuffle.method": "p2p"}):
-        df.shuffle("x")
+    if dd._dask_expr_enabled():
+        df.shuffle("x").optimize()
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_graph.py:75]: From what I understand, this should raise.

File: distributed/shuffle/tests/test_graph.py [hunk 4]
```diff
@@ -65,18 +71,25 @@ def test_raise_on_sparse_data():
...
 
 def test_raise_on_non_string_column_name():
     df = dd.from_pandas(pd.DataFrame({"a": range(10), 1: range(10)}), npartitions=5)
-    with pytest.raises(
-        TypeError, match="p2p requires all column names to be str"
-    ), dask.config.set({"dataframe.shuffle.method": "p2p"}):
-        df.shuffle("a")
+
+    if dd._dask_expr_enabled():
+        df.shuffle("a").optimize()
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_graph.py:87]: Why does this not raise?

File: distributed/shuffle/tests/test_merge.py [hunk 5]
```diff
@@ -81,30 +80,27 @@ async def test_basic_merge(c, s, a, b, how):
     B = pd.DataFrame({"y": [1, 3, 4, 4, 5, 6], "z": [6, 5, 4, 3, 2, 1]})
     b = dd.repartition(B, [0, 2, 5])
 
-    joined = hash_join(a, "y", b, "y", how)
+    joined = a.merge(b, left_on="y", right_on="y", how=how)
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_merge.py:83]: We should ensure that we indeed use a hashjoin, otherwise this test is worthless.

File: distributed/shuffle/tests/test_merge.py [hunk 6]
```diff
@@ -222,10 +218,8 @@ async def test_merge(c, s, a, b, how, disk):
             await c.compute(dd.merge(A, b, how=how)),
             pd.merge(A, B, how=how),
         )
-        # Note: No await since A and B are both pandas dataframes and this doesn't
-        # actually submit anything
         list_eq(
-            c.compute(dd.merge(A, B, how=how)),
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_merge.py:228]: It looks like this wouldn't work on "legacy" `dask.dataframe` anymore. Is this intended?

File: distributed/tests/test_worker.py [hunk 7]
```diff
@@ -3446,6 +3446,9 @@ def get_data(self, comm, **kwargs):
         return super().get_data(comm, **kwargs)
 
 
+@pytest.mark.xfail(
+    reason="https://github.com/dask-contrib/dask-expr/issues/945", raises=UserWarning
+)
```
Comment (crusaderky) [inline distributed/tests/test_worker.py:3451]: Not too happy about xfailing this, as it's not testing anything dataframe related.
I'll see if I can rewrite it with a generic graph.

Comment (milesgranger) [reply distributed/tests/test_worker.py:3451]: Doesn't specifying the raises keyword mean it's only failed if that warning is raised but would hard fail otherwise, is that right?

File: distributed/shuffle/tests/test_graph.py (hunk 4 above)
Comment (milesgranger) [reply distributed/shuffle/tests/test_graph.py:82]: It's still a TODO to make an issue for this and the others to raise properly but haven't gotten to it yet.

File: distributed/shuffle/tests/test_graph.py (hunk 2 above)
Comment (milesgranger) [reply distributed/shuffle/tests/test_graph.py:38]: I agree, this should also raise properly. Need to open an issue in dask-expr with this and the other raises tests here.

File: distributed/tests/test_worker.py (hunk 7 above)
Comment (phofl) [reply distributed/tests/test_worker.py:3451]: Can we just catch the warning? No need to fail the whole test

File: distributed/shuffle/tests/test_graph.py (hunk 3 above)
Comment (milesgranger) [reply distributed/shuffle/tests/test_graph.py:70]: Yep, and this is the other. Maybe xfailing these is better until updated in dask-expr. Looking at them again now like this seems not great.

File: distributed/tests/test_worker.py (hunk 7 above)
Comment (milesgranger) [reply distributed/tests/test_worker.py:3451]: ya, that's probably better, and actually, I think we can just remove this b/c in a later commit that was squashed into this, I just added the ignore into `pyproject.toml` (https://github.com/dask/distributed/pull/8560/commits/f71ea01ead8f7f3dd55839440aa2e0b1bfe20e34)

Comment (milesgranger) [reply distributed/tests/test_worker.py:3451]: Okay done.. https://github.com/dask/distributed/pull/8560/commits/df07359e1e4834b4859c3fd830539045f094a593, sorry for the confusion. :sweat_smile:

File: distributed/shuffle/tests/test_graph.py (hunk 2 above)
Comment (phofl) [reply distributed/shuffle/tests/test_graph.py:38]: https://github.com/dask-contrib/dask-expr/pull/951

File: distributed/shuffle/tests/test_graph.py (hunk 4 above)
Comment (phofl) [reply distributed/shuffle/tests/test_graph.py:82]: https://github.com/dask-contrib/dask-expr/pull/951

File: distributed/shuffle/tests/test_graph.py (hunk 3 above)
Comment (phofl) [reply distributed/shuffle/tests/test_graph.py:70]: https://github.com/dask-contrib/dask-expr/pull/951

File: distributed/shuffle/tests/test_merge.py (hunk 6 above)
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_merge.py:228]: I've confirmed that this raises with `TypeError: object DataFrame can't be used in 'await' expression`.

File: distributed/shuffle/tests/test_merge.py (hunk 5 above)
Comment (milesgranger) [reply distributed/shuffle/tests/test_merge.py:83]: Looks like here it's a `Merge`, is that okay?
```python
(Pdb) p list(joined.optimize().walk())
[Merge(4266cea), df, df]
```

Comment (phofl) [reply distributed/shuffle/tests/test_merge.py:83]: No, only ``HashJoinP2P`` is equivalent to hash_join, but it might just be a repr thing. I would assert explicitly

File: distributed/shuffle/tests/test_merge.py (hunk 6 above)
Comment (phofl) [reply distributed/shuffle/tests/test_merge.py:228]: I think I avoided that on purpose. But I don't have a very strong opinion either way. The thing I didn't like about this is that we return a pandas DataFrame instead of a Dask DataFrame In this case

File: distributed/shuffle/tests/test_merge.py (hunk 5 above)
Comment (milesgranger) [reply distributed/shuffle/tests/test_merge.py:83]: eh, not looking good, this fails:
```python
    from dask_expr._merge import HashJoinP2P
    assert any(isinstance(expr, HashJoinP2P) for expr in joined._expr.walk())
```
Looks like it's `Merge`.

```python
(Pdb) p list(joined._expr.walk())
[Merge(4fd85ee), df, df]
(Pdb) p type(list(joined._expr.walk())[0])
<class 'dask_expr._merge.Merge'>
```

Comment (phofl) [reply distributed/shuffle/tests/test_merge.py:83]: you have to call optimise before it can be a HashJoin

Comment (milesgranger) [reply distributed/shuffle/tests/test_merge.py:83]: :man_facepalming: ah, right. (https://github.com/dask/distributed/pull/8560/commits/a67ac1d6d4e8049364f3c7cac49870fbbd76120a)

File: distributed/shuffle/tests/test_merge.py (hunk 6 above)
Comment (milesgranger) [reply distributed/shuffle/tests/test_merge.py:228]: Super, first thought is I like receiving dask dataframes when I use merge from dask, so made it conditional in https://github.com/dask/distributed/pull/8560/commits/a165f1601de7bec2637cd8c4cf458f6f91f0f2fc, but happy to adapt.

File: distributed/protocol/tests/test_highlevelgraph.py (hunk 1 above)
Comment (milesgranger) [reply distributed/protocol/tests/test_highlevelgraph.py]: Okay, updated to check for that warning instead now. https://github.com/dask/distributed/pull/8560/commits/302b8bb0e9d614a79817c1c7084d197acb891e24

File: distributed/tests/test_dask_collections.py [hunk 8]
```diff
@@ -41,6 +43,9 @@ def assert_equal(a, b):
         assert a == b
 
 
+@pytest.mark.xfail(
+    dd._dask_expr_enabled(), reason="P2P transfer error", raises=RuntimeError
```
Comment (hendrikmakait) [inline distributed/tests/test_dask_collections.py:47]: This is https://github.com/dask-contrib/dask-expr/issues/955

Comment (milesgranger) [reply distributed/tests/test_dask_collections.py]: thanks! https://github.com/dask/distributed/pull/8560/commits/92204a5feb31324b65f83e8bfb8f582ee0abde20

Comment (hendrikmakait) [reply distributed/tests/test_dask_collections.py]: Actually, the issue is slightly different, I'll provide a fix.

File: distributed/tests/test_scheduler.py [hunk 9]
```diff
@@ -2763,13 +2763,16 @@ async def test_get_task_duration(c, s, a, b):
...
-    """Ensure that the default task durations for shuffle split tasks are, by default,
+    """Ensure that the default task durations for shuffle split tasks are
     aligned with the task names of dask.dask
     """
     pd = pytest.importorskip("pandas")
     dd = pytest.importorskip("dask.dataframe")
 
-    # We don't care about the actual computation here but we'll schedule one anyhow to
+    if dd._dask_expr_enabled():
+        pytest.skip("No split task in dask-expr")
```
Comment (crusaderky) [inline distributed/tests/test_scheduler.py:2773]: @fjetter @phofl can I have a secondo opinion on this? Not sure I understand why we don't have this task in dask-expr. Are we missing a line in `distributed.scheduler.default-task-durations`?

Comment (crusaderky) [reply distributed/tests/test_scheduler.py:2773]: Same discussion for `test_blocklist_shuffle_split`

File: distributed/tests/test_steal.py [hunk 10]
```diff
@@ -1011,6 +1011,7 @@ async def test_balance_with_longer_task(c, s, a, b):
     assert z.key in b.data
 
 
+@pytest.mark.xfail(reason="blocked is empty", raises=AssertionError)
```
Comment (crusaderky) [inline distributed/tests/test_steal.py:1014]: Potential blocker.
Discussion on `test_scheduler.py::test_default_task_duration_splits` above.

File: distributed/tests/test_scheduler.py (hunk 9 above)
Comment (phofl) [reply distributed/tests/test_scheduler.py:2773]: @hendrikmakait and I stumbled upon the root cause, dask-expr doesn't properly respect the context manager. I am preparing a fix at the moment, should be there in a minute

Comment (phofl) [reply distributed/tests/test_scheduler.py:2773]: https://github.com/dask-contrib/dask-expr/pull/958

Comment (phofl) [reply distributed/tests/test_scheduler.py:2773]: pr is merged

File: distributed/shuffle/tests/test_shuffle.py [hunk 11]
```diff
@@ -1340,50 +1340,44 @@ async def test_repeat_shuffle_operation(c, s, a, b, wait_until_forgotten):
...
+    out = c.submit(
+        block, out, in_event, block_event, workers=[b.address], allow_other_workers=True
+    )
 
-        await c.close()
-        await check_worker_cleanup(a)
-        await check_scheduler_cleanup(s)
+    await wait_until_worker_has_tasks(UNPACK_PREFIX, b.address, 1, s)
+    await in_event.wait()
+    await s.remove_worker(b.address, stimulus_id="remove_b")
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:1370]: Note that this tests a different error condition than the Nanny-based version of the test. I _think_ this error condition is valid, but I'd rather extract this into a separate PR to be able to think this through.

File: distributed/tests/test_scheduler.py (hunk 9 above)
Comment (crusaderky) [reply distributed/tests/test_scheduler.py:2773]: @phofl I'm now seeing `split-taskshuffle` and `split-stage`. Are they both supposed to have 1us default-duration?

File: distributed/shuffle/tests/test_shuffle.py (hunk 11 above)
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_shuffle.py:1370]: @crusaderky: Is there a specific reason to have https://github.com/dask/distributed/pull/8560/commits/9ab8ca6f863092775c5676c6ebf6e6398b9d0847 in this PR?

Comment (crusaderky) [reply distributed/shuffle/tests/test_shuffle.py:1370]: @hendrikmakait the test needed refactoring to use submit instead of delayed due to annotations.
I can try to minimize the change here and make the more involved one in a separate PR.

File: distributed/tests/test_scheduler.py (hunk 9 above)
Comment (phofl) [reply distributed/tests/test_scheduler.py]: Yeah the naming is confusing and something we probably should fix, but it's ok conceptually

**Comment by crusaderky** [comment]
I'm aiming to merge this PR with these tests still failing, as they are more involved:
- test_dataframe_set_index_sync (#8561)
- test_default_task_duration_splits (default durations)
- test_blocklist_shuffle_split (default durations)

File: distributed/shuffle/tests/test_shuffle.py (hunk 11 above)
Comment (crusaderky) [reply distributed/shuffle/tests/test_shuffle.py]: Done

**Review by hendrikmakait — APPROVED** [review]
Thanks, everyone! I agree with @crusaderky that we should go ahead and merge if nothing else pops up on CI.

**Review by crusaderky — APPROVED** [review]
